xbuild/src/cargo/mod.rs

```diff
@@ -24,11 +30,60 @@ impl Cargo {
         target_dir: Option<PathBuf>,
         offline: bool,
     ) -> Result<Self> {
-        let (manifest, package) = utils::find_package(
-            &manifest_path.unwrap_or_else(|| std::env::current_dir().unwrap()),
-            package,
-        )?;
-        let root_dir = manifest.parent().unwrap();
+        let manifest_path = manifest_path
+            .map(|path| {
+                if path.file_name() != Some(OsStr::new("Cargo.toml")) || !path.is_file() {
+                    Err(anyhow::anyhow!(
+                        "The manifest-path must be a path to a Cargo.toml file"
+                    ))
+                } else {
+                    Ok(path)
+                }
+            })
+            .transpose()?;
+
+        let search_path = manifest_path.map_or_else(
+            || std::env::current_dir().unwrap(),
+            |manifest_path| manifest_path.parent().unwrap().to_owned(),
+        );
+
+        // Scan the given and all parent directories for a Cargo.toml containing a workspace
+        let workspace_manifest = utils::find_workspace(&search_path)?;
+
+        let (manifest_path, manifest) =
+            if let (Some(package), Some((workspace_manifest_path, workspace))) =
+                (package, &workspace_manifest)
+            {
+                // If a workspace was found, and the user chose a package with `-p`, find packages relative to it
+                // TODO: What if we call `cargo apk run` in the workspace root, and detect a workspace? It should
+                // then use the `[package]` defined in the workspace (will be found below, though, but currently
+                // fails with UnexpectedWorkspace)
+                utils::find_package_manifest_in_workspace(
```

Comment:
Most notably this TODO: https://github.com/dvc94ch/cargo-subcommand/pull/25#discussion_r1032611548

Reply:
@dvc94ch before re-approving: this is the only thing that really needs to be looked into....

Reply:
@dvc94ch I ended up implementing this (and some more fixes) in https://github.com/dvc94ch/cargo-subcommand/pull/25, can you please take a look at the last 3 commits? If you are okay with that approach I'll transplant the changes to `xbuild`!